File: tools/check_gemini_generate.py
#!/usr/bin/env python3
import json
import urllib.error
import urllib.request


def load_env_key(path: str, name: str) -> str | None:
    try:
        with open(path, "r", encoding="utf-8") as f:
            for raw in f:
                line = raw.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                k, v = line.split("=", 1)
                if k.strip() != name:
                    continue
                return v.strip().strip('"').strip("'")
    except FileNotFoundError:
        return None


def main() -> None:
    key = load_env_key(".env", "GEMINI_API_KEY")
    if not key:
        raise SystemExit("No GEMINI_API_KEY found in .env")

    url = (
        "https://generativelanguage.googleapis.com/v1beta/"
        "models/gemini-3-pro-preview:generateContent"
        f"?key={key}"
    )
    payload = {
        "contents": [
            {"role": "user", "parts": [{"text": "Reply with exactly: ping"}]}
        ],
        "generationConfig": {"temperature": 0.2, "maxOutputTokens": 16},
    }

    req = urllib.request.Request(
        url,
        data=json.dumps(payload).encode("utf-8"),
        headers={"Content-Type": "application/json"},
    )

    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            body = resp.read().decode("utf-8", "replace")
            print("HTTP", resp.status)
            data = json.loads(body)
            parts = (
                (data.get("candidates") or [{}])[0]
                .get("content", {})
                .get("parts")
                or []
            )
            text = "".join([p.get("text", "") for p in parts]).strip()
            print("text=", text)
            return
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", "replace")
        print("HTTP", e.code)
        try:
            j = json.loads(body)
            err = j.get("error", {})
            print("status=", err.get("status"))
            msg = str(err.get("message") or "")
            print("message(first 240)=", msg[:240])
        except Exception:
            print("body(first 240)=", body[:240])


if __name__ == "__main__":
    main()

File: tools/check_gemini_generate.js
#!/usr/bin/env node

const fs = require('fs');

function loadEnvKey(path, name) {
  try {
    const text = fs.readFileSync(path, 'utf8');
    for (const raw of text.split(/\r?\n/)) {
      const line = raw.trim();
      if (!line || line.startsWith('#') || !line.includes('=')) continue;
      const idx = line.indexOf('=');
      const k = line.slice(0, idx).trim();
      if (k !== name) continue;
      return line
        .slice(idx + 1)
        .trim()
        .replace(/^['"]|['"]$/g, '');
    }
  } catch {
    // ignore
  }
  return null;
}

async function main() {
  const key = loadEnvKey('.env', 'GEMINI_API_KEY');
  if (!key) {
    throw new Error('No GEMINI_API_KEY found in .env');
  }

  const url =
    'https://generativelanguage.googleapis.com/v1beta/' +
    'models/gemini-3-pro-preview:generateContent' +
    `?key=${encodeURIComponent(key)}`;

  const payload = {
    contents: [{ role: 'user', parts: [{ text: 'Reply with exactly: ping' }] }],
    generationConfig: { temperature: 0.2, maxOutputTokens: 256 },
  };

  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });

  const bodyText = await res.text();
  console.log('HTTP', res.status);

  let body;
  try {
    body = JSON.parse(bodyText);
  } catch {
    body = null;
  }

  if (!res.ok) {
    const err = body?.error;
    console.log('status=', err?.status);
    const msg = String(err?.message || bodyText || '');
    console.log('message(first 240)=', msg.slice(0, 240));
    process.exit(1);
  }

  const cand = body?.candidates?.[0];
  const parts = cand?.content?.parts || [];
  const text = parts.map((p) => p?.text || '').join('').trim();
  console.log('text=', text);
  if (!text) {
    console.log('finishReason=', cand?.finishReason);
    console.log('hasParts=', Array.isArray(parts) ? parts.length : 0);
    const usage = body?.usageMetadata;
    if (usage) {
      console.log('usageMetadata=', JSON.stringify(usage));
    }
    console.log('raw(first 800)=', JSON.stringify(body).slice(0, 800));
    const safety = cand?.safetyRatings;
    if (safety) {
      console.log('safetyRatings=', JSON.stringify(safety).slice(0, 240));
    }
    const promptFeedback = body?.promptFeedback;
    if (promptFeedback) {
      console.log('promptFeedback=', JSON.stringify(promptFeedback).slice(0, 240));
    }
  }
}

main().catch((err) => {
  console.error(err?.message || String(err));
  process.exit(1);
});
